runtime/compiler/env/j9methodServer.cpp

```diff
@@ -351,54 +357,36 @@ TR_ResolvedJ9JITServerMethod::getResolvedPossiblyPrivateVirtualMethod(TR::Compil
          !comp->ilGenRequest().details().isMethodHandleThunk() && // cmvc 195373
          performTransformation(comp, "Setting as unresolved virtual call cpIndex=%d\n",cpIndex) ) || ignoreRtResolve || shouldCompileTimeResolve)
       {
-      _stream->write(JITServer::MessageType::ResolvedMethod_getResolvedPossiblyPrivateVirtualMethodAndMirror, (TR_ResolvedMethod *) _remoteMirror, literals(), cpIndex);
-      auto recv = _stream->read<J9Method *, UDATA, bool, TR_ResolvedJ9JITServerMethodInfo>();
-      J9Method *ramMethod = std::get<0>(recv);
-      UDATA vTableIndex = std::get<1>(recv);
-      bool isUnresolvedInCP = std::get<2>(recv);
+      _stream->write(JITServer::MessageType::ResolvedMethod_getResolvedPossiblyPrivateVirtualMethodAndMirror, _remoteMirror, literals(), cpIndex, ignoreRtResolve);
+      auto recv = _stream->read<TR_OpaqueMethodBlock *, TR_ResolvedJ9JITServerMethodInfo, uint32_t, bool>();
+      TR_OpaqueMethodBlock *ramMethod = std::get<0>(recv);
+      auto methodInfo = std::get<1>(recv);
+      uint32_t vTableSlot = std::get<2>(recv);
+      bool isUnresolvedInCP = std::get<3>(recv);
 
       if (unresolvedInCP)
          *unresolvedInCP = isUnresolvedInCP;
 
       bool createResolvedMethod = true;
 
-      if (comp->compileRelocatableCode() && ramMethod && comp->getOption(TR_UseSymbolValidationManager))
-         {
-         if (!comp->getSymbolValidationManager()->addVirtualMethodFromCPRecord((TR_OpaqueMethodBlock *)ramMethod, cp(), cpIndex))
-            createResolvedMethod = false;
-         }
-
-      if (vTableIndex)
+      if (ramMethod == NULL) 
          {
-         TR_AOTInliningStats *aotStats = NULL;
-         if (comp->getOption(TR_EnableAOTStats))
-            aotStats = & (((TR_JitPrivateConfig *)_fe->_jitConfig->privateConfig)->aotStats->virtualMethods);
-
-         TR_ResolvedJ9JITServerMethodInfo &methodInfo = std::get<3>(recv);
-         
-         // call constructor without making a new query
-         if (createResolvedMethod)
+         if (unresolvedInCP) 
             {
-            resolvedMethod = createResolvedMethodFromJ9Method(comp, cpIndex, vTableIndex, ramMethod, unresolvedInCP, aotStats, methodInfo);
-            compInfoPT->cacheResolvedMethod(compInfoPT->getResolvedMethodKey(TR_ResolvedMethodType::VirtualFromCP, (TR_OpaqueClassBlock *) _ramClass, cpIndex), (TR_OpaqueMethodBlock *) ramMethod, (uint32_t) vTableIndex, methodInfo);
+            handleUnresolvedVirtualMethodInCP(cpIndex, unresolvedInCP);
             }
+         return NULL;
          }
+      
+      bool success;
+      compInfoPT->cacheResolvedMethod(
+         key,
+         ramMethod,
+         vTableSlot,
+         methodInfo);
+      success = compInfoPT->getCachedResolvedMethod(key, this, &resolvedMethod, unresolvedInCP);
+      
```

Comment:
Since this sequence of method calls is used frequently now, I think it'd make sense to make it into a helper function, and call it something like `cacheAndGetResolvedMethod`. The helper method would have to go into `TR::CompilationInfoPerThreadRemote`.